contracts/swap/swap.go

```diff
@@ -0,0 +1,171 @@
+
+// Package swap wraps the 'swap' Ethereum smart contract.
+// It is an abstraction layer to hide implementation details about the different
+// Swap contract iterations (Simple Swap, Soft Swap, etc.)
+package swap
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethersphere/swarm/contracts/swap/contract"
+)
+
+var (
+	// ErrNotASwapContract is given when an address is verified not to have a SWAP contract based on its bytecode
+	ErrNotASwapContract = errors.New("not a swap contract")
+	// ErrTransactionReverted is given when the transaction that submits or cashes a cheque is reverted
+	ErrTransactionReverted = errors.New("Transaction reverted")
+)
+
+// Backend wraps all methods required for contract deployment.
+type Backend interface {
+	bind.ContractBackend
+	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+}
+
+// Deploy deploys an instance of the underlying contract and returns its `Contract` abstraction
+func Deploy(auth *bind.TransactOpts, backend bind.ContractBackend, owner common.Address, harddepositTimeout time.Duration) (common.Address, Contract, *types.Transaction, error) {
+	addr, tx, s, err := contract.DeploySimpleSwap(auth, backend, owner, big.NewInt(int64(harddepositTimeout)))
+	c := simpleContract{instance: s}
+	return addr, c, tx, err
+}
+
+// InstanceAt creates a new instance of a contract at a specific address.
+// It assumes that there is an existing contract instance at the given address, or an error is returned
+// This function is needed to communicate with remote Swap contracts (e.g. sending a cheque)
+func InstanceAt(address common.Address, backend bind.ContractBackend) (Contract, error) {
+	simple, err := contract.NewSimpleSwap(address, backend)
+	c := simpleContract{instance: simple}
+	return c, err
+}
+
+// Contract interface defines the methods exported from the underlying go-bindings for the smart contract
+type Contract interface {
+	// Submit a cheque to the beneficiary
+	SubmitChequeBeneficiary(opts *bind.TransactOpts, backend Backend, serial *big.Int, amount *big.Int, timeout *big.Int, ownerSig []byte) (*types.Receipt, error)
+	// Cash the cheque by the beneficiary
+	CashChequeBeneficiary(auth *bind.TransactOpts, backend Backend, beneficiary common.Address, requestPayout *big.Int) (*types.Receipt, error)
+	// Return contract info (e.g. deployed address)
+	ContractParams() *Params
+	// Return the contract owner from the blockchain
+	Issuer(opts *bind.CallOpts) (common.Address, error)
+	// Return the last cheque
+	Cheques(opts *bind.CallOpts, addr common.Address) (*ChequeResult, error)
+}
+
+// ChequeResult is needed because the underlying `Cheques` method returns an untyped struct
+type ChequeResult struct {
+	Serial      *big.Int
```

Comment:
Similarly to above there is not really an alternative, but since this struct is filled from return data from a smart contract interaction we at least no that all those `big.Int` are always less than 2^256.